_posts/2015-09-01-Cocoa-Architecture-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,148 @@
+
+Network models also make it extremely easy to swap behavior out in tests. In eigen, we have our asynchronous networking [run synchronously in tests](https://github.com/artsy/eigen/pull/575) but we still use the network models to be able to provide [whatever data we want to expect](https://github.com/artsy/eigen/blob/master/Artsy_Tests/View_Controller_Tests/Artist/ARArtistViewControllerTests.m#L29-L40) from the server in our tests.
+
+## Subclassing more than twice
+
+As projects evolve it can become very easy to subclass _x_ in order to provide a "similar but a little bit different" behavior. Perhaps you need to [override some methods](https://github.com/artsy/eigen/blob/e19ac594bf6240d076e8092d9c56e9876c94444e/Artsy/View_Controllers/Web_Browsing/ARTopMenuInternalMobileWebViewController.m#L58), or add a [specific behavior](https://github.com/artsy/eigen/blob/e19ac594bf6240d076e8092d9c56e9876c94444e/Artsy/View_Controllers/Web_Browsing/AREndOfLineInternalMobileWebViewController.h#L5). Like the [urban myth](http://ezinearticles.com/?The-Boiled-Frog-Phenomenon&id=932310) of a frog being slowly boiled, you end up with a difficult to understand codebase as expected behavior mutates depending on how deep the hierarchy goes.
+
+One pattern for handling this is [class composition](http://stackoverflow.com/questions/9710411/ios-grasping-composition). This is the idea that instead of having one object do multiple things, you allow a collection of objects to work together. Providing more space for each object to conform to the Single Responsibility Principal ([SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle).) If you're interested in this, you may also be interested in the [class clusters](https://developer.apple.com/library/ios/documentation/General/Conceptual/CocoaEncyclopedia/ClassClusters/ClassClusters.html) pattern.
+
+A good example of this comes [from Energy](https://github.com/artsy/energy/blob/aa97d90cf37932d4c0f49ea4c4d31f7e491f16a6/Classes/Controllers/Top%20View%20Controller/ARTopViewToolbarController.m), our root view controller `ARTopViewController` used to control its own toolbar items. Over 4 years this became difficult to manage, and a lot of extra code in the view controller. By abstracting out the implementation details of managing the toolbar items into it's [own class](https://github.com/artsy/energy/blob/aa97d90cf37932d4c0f49ea4c4d31f7e491f16a6/Classes/Controllers/Top%20View%20Controller/ARTopViewToolbarController.m) we were able to allow the `ARTopViewController` to state what it wanted by not how it was done.
```

Comment:
> we were able to allow the `ARTopViewController` to state what it wanted ~~by~~ **instead of** how it was done.

I don’t know for sure what the last part of that sentence is supposed to be, is it this?